File: 0104-maximum-depth-of-binary-tree/0104-maximum-depth-of-binary-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func maxDepth(root *TreeNode) int {
    if root == nil {
        return 0
    }
    
    leftCount := solution(root.Left, 1)
    rightCount := solution(root.Right, 1)
    
    if leftCount > rightCount {
        return leftCount
    }
    
    return rightCount
}

func solution(node *TreeNode, count int) int {
    if node == nil {
        return count
    }
    
    leftCount := solution(node.Left, count + 1)
    rightCount := solution(node.Right, count + 1)
    
    if leftCount > rightCount {
        return leftCount
    }
    
    return rightCount
}

File: 0104-maximum-depth-of-binary-tree/0104-maximum-depth-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int maxDepth(TreeNode root) {
        if (root == null) {
            return 0;
        }
        
        return Math.max(solution(root.left, 1), solution(root.right, 1));
    }
    
    private int solution(TreeNode node, int count) {
        if (node == null) {
            return count;
        }
        
        return Math.max(solution(node.left, count + 1), solution(node.right, count + 1));
    }
}